ml-agents/mlagents/trainers/buffer.py

```diff
@@ -12,321 +13,260 @@ class BufferException(UnityException):
+    def truncate(self, max_length: int, sequence_length: int = 1) -> None:
         """
-        Truncates the update buffer to a certain length.
+        Truncates the buffer to a certain length.
 
         This can be slow for large buffers. We compensate by cutting further than we need to, so that
         we're not truncating at each update. Note that we must truncate an integer number of sequence_lengths
         param: max_length: The length at which to truncate the buffer.
         """
-        current_length = len(next(iter(self.update_buffer.values())))
+        current_length = self.num_experiences
         # make max_length an integer number of sequence_lengths
         max_length -= max_length % sequence_length
         if current_length > max_length:
-            for _key in self.update_buffer.keys():
-                self.update_buffer[_key] = self.update_buffer[_key][
-                    current_length - max_length :
-                ]
+            for _key in self.keys():
+                self[_key] = self[_key][current_length - max_length :]
 
-    def reset_local_buffers(self):
+    @property
+    def num_experiences(self):
         """
-        Resets all the local local_buffers
-        """
-        agent_ids = list(self.keys())
-        for k in agent_ids:
-            self[k].reset_agent()
+        The number of agent experiences in the AgentBuffer, i.e. the length of the buffer.
 
-    def append_update_buffer(
-        self, agent_id, key_list=None, batch_size=None, training_length=None
-    ):
-        """
-        Appends the buffer of an agent to the update buffer.
-        :param agent_id: The id of the agent which data will be appended
-        :param key_list: The fields that must be added. If None: all fields will be appended.
-        :param batch_size: The number of elements that must be appended. If None: All of them will be.
-        :param training_length: The length of the samples that must be appended. If None: only takes one element.
+        An experience consists of one element across all of the fields of this AgentBuffer.
+        Note that these all have to be the same length, otherwise shuffle and append_to_update_buffer
+        will fail.
         """
-        if key_list is None:
-            key_list = self[agent_id].keys()
-        if not self[agent_id].check_length(key_list):
+        key_list = list(self.keys())
+        if not self.check_length(key_list):
             raise BufferException(
-                "The length of the fields {0} for agent {1} where not of same length".format(
-                    key_list, agent_id
-                )
-            )
-        for field_key in key_list:
-            self.update_buffer[field_key].extend(
-                self[agent_id][field_key].get_batch(
-                    batch_size=batch_size, training_length=training_length
-                )
+                "Unable to shuffle if the fields are not of same length"
```

Comment:
This error message doesn't necessarily make sense here (you might not be shuffling at the time).

Reply:
Oh no, I left some test code in... checking the lengths of the fields every time this was called made things significantly slower, so I removed it.